plugin/SqueezeESP32/PlayerSettings.pm

```diff
@@ -30,63 +30,60 @@ sub page {
 
 sub prefs {
 	my ($class, $client) = @_;
-	my @prefs = qw(width small_VU spectrum artwork eq);
+	my @prefs;
+	push @prefs, qw(width small_VU) if $client->displayWidth;
 	return ($prefs->client($client), @prefs);
```

Comment:
So we don't need to set other prefs (the non-scalar ones) here as we do that in the handler?

Reply:
Exactly. Actually you don't want them to be handled by the super class, as it might break things. As you've seen you had to read the prefs yourself anyway, because `Settings` wouldn't treat them correctly. Better take them in your hands and don't let that code touch them.

I never really looked into that code. But I must assume that LMS itself doesn't have many "complex" prefs. Or didn't back in the days. The base class is/was good for 9x% of the code, thus never enhanced this way.

Reply:
understood - thanks for the explanation

Reply:
I did some test this week-end as I was pushing the IR changes and if I don't set in @prefs  spectrum, artwork,and eq then there is nothing set in setting screen for these items. Then what is saved is 0/no value and that causes a div0 when updating. 
As said, I don't understand well this part of Settings, so I need to have a better look. It this an issue where beforeHandler would be required to set the prefs?

Reply:
Oh, I believe I see what is missing: the template is accessing `prefs.pref_spectrum`, but you store the prefs in `pref_spectrum` etc. directly. I'll try to fix that.

Reply:
Background being that the `Settings` base class would store prefs in the `prefs` element to make sure there are no naming conflicts with other items in the `$paramRef`.

Reply:
See https://github.com/sle118/squeezelite-esp32/pull/27

Reply:
ok, I think I'm starting so understand this settings things. So every 'xxx' that is set through the "prefs" method will end up being stored in a "prefs." element, but not as prefs.xxx but as prefs.pref_xxx. But in paramsRef, they should still be addressed as paramsRef->{'ref_xxx'}.

Although, if you are not settings xxx in 'prefs' method, you can still use the paramsRef->
{'ref_xxx'} if you want to, but using the 'pref_' prefix is really useless, in my example I could have used 'spectrum' directly, it's just a better practice to avoid name conflict.

I guess what always confused me is the magic that the class does when a paramRef hash 'ref_xxx' exists *and* 'xxx' has been set in the 'prefs' method.

Reply:
No, in `$paramsRef` they should still be addressed as `$paramsRef->{prefs}->{pref_xxx}`. But the thing is: if I want to deal with a pref in my code, then I'm not putting it on the list returned by the `prefs()` method. And in that case I can call it whatever I want, as long as I don't create a conflict. You'd also have to handle the pref yourself if it wasn't a simple scalar value. In that case, again, you can call it whatever you want. 

The issue I had created was that I _removed_ some prefs from the list dealt with by the super class, but the template relied on them being stored in `prefs.pref_xxx`. At the same time you set `$paramRef->{pref_xxx}`, which actually resulted in those prefs to be in `$paramRef` twice, once within the `prefs` attribute, and once outside. You modified and stored the latter, but the template read from the former. After I removed those prefs from the list dealt with automatically, `prefs.pref_xxx` was no longer set.

Reply:
understood I think. What confused me as well was that although I was setting paramRef->{pref_xxx} and getting the result from these as well (under savesettings), the access in the html was using prefs.pref_xxx. I assume that worked because they were both a reference to the same object, not copies. 
Similarly as the template toolkit let you use things like
``` 
	[% WRAPPER setting title="PLUGIN_SQUEEZEESP32_WIDTH" desc="PLUGIN_SQUEEZEESP32_WIDTH_DESC" %]
			<input type="hidden" name="pref_width" value="[% prefs.pref_width %]">
			[% prefs.pref_width %]
		[% END %]
```
Which got me confused about when the "pref_width" is "created" because I don't create it, it's only set in the methods "prefs". I think in the future I wont prefix my own (non-auto-managed) preferences by 'pref_' but by something else so that I don't forget. 
There are probably too much C muscle thinking in me (or not enough muscles :-)